contentcuration/contentcuration/utils/publish.py

```diff
@@ -808,6 +815,54 @@ def fill_published_fields(channel, version_notes):
     channel.save()
 
 
+def sync_contentnode_and_channel_tsvectors(channel_id):
+    """
+    Creates, deletes and updates tsvectors of the channel and all its content nodes
+    to reflect the current state of channel's main tree.
+    """
+    # Update or create channel tsvector entry.
+    logging.info("Setting tsvector for channel with id {}.".format(channel_id))
+
+    channel = (get_fts_annotated_channel_qs()
+               .values("keywords_tsvector", "main_tree__tree_id")
+               .get(pk=channel_id))
+
+    obj, is_created = ChannelFullTextSearch.objects.update_or_create(channel_id=channel_id, defaults={"keywords_tsvector": channel["keywords_tsvector"]})
+    del obj
+
+    if is_created:
+        logging.info("Created 1 channel tsvector.")
+    else:
+        logging.info("Updated 1 channel tsvector.")
+
+    # Update or create contentnodes tsvector entry for channel_id.
+    logging.info("Setting tsvectors for all main tree contentnodes in channel {}.".format(channel_id))
+
+    if ContentNodeFullTextSearch.objects.filter(channel_id=channel_id).exists():
+        # First, delete nodes that are no longer in main_tree.
+        nodes_no_longer_in_main_tree = ~Exists(ccmodels.ContentNode.objects.filter(id=OuterRef("contentnode_id"), tree_id=channel["main_tree__tree_id"]))
+        ContentNodeFullTextSearch.objects.filter(nodes_no_longer_in_main_tree, channel_id=channel_id).delete()
+
```

Comment:
I had started a review, but this was my only real comment so far - it's not a blocker, but more that I would really prefer that we reference Python imports directly rather than using `call_command`, and we should consolidate these sorts of utilities somewhere else and then reference them in management commands.

I feel like it would be simpler to do a bulk create for the new nodes here, rather than then deferring to the management command below? If you create the objects here, then the update statement below does all the updates in a single query.

If you want to share functionality between the management command and the publish utils here, I'd either a standalone utility function for some of this, or add it to the manager.

Reply:
Resolved on slack.